src/httpClient/default.ts

```diff
@@ -0,0 +1,220 @@
+  private baseUrl: string;
+  private origin: URL;
+  private apiKey: string;
+  private connectTimeout: number;
+  readonly userAgent = `tpuf-typescript/${version}`;
+  private compression: boolean;
+
+  constructor(
+    baseUrl: string,
+    apiKey: string,
+    connectTimeout: number,
+    idleTimeout: number, // not used currently, just to match NodeHTTPClient's signature
+    warmConnections: number,
+    compression: boolean,
+  ) {
+    this.baseUrl = baseUrl;
+    this.origin = new URL(baseUrl);
+    this.origin.pathname = "";
+    this.apiKey = apiKey;
+    this.connectTimeout = connectTimeout;
+    this.compression = compression;
+
+    for (let i = 0; i < warmConnections; i++) {
+      // send a small request to put some connections in the pool
+      void fetch(this.baseUrl, {
+        method: "HEAD",
+        headers: { "User-Agent": this.userAgent },
+      });
+    }
+  }
+
+  async doRequest<T>({
+    method,
+    path,
+    query,
+    body,
+    compress,
+    retryable,
+  }: RequestParams): RequestResponse<T> {
+    const url = new URL(`${this.baseUrl}${path}`);
+    if (query) {
+      Object.keys(query).forEach((key) => {
+        const value = query[key];
+        if (value) {
+          url.searchParams.append(key, value);
+        }
+      });
+    }
+    path = url.pathname;
+    if (query) {
+      path += "?" + url.search;
+    }
+
+    const headers: Record<string, string> = {
+      Authorization: `Bearer ${this.apiKey}`,
+      "User-Agent": this.userAgent,
+    };
+
+    if (this.compression) {
+      headers["Accept-Encoding"] = "gzip";
+    }
+
+    if (body) {
+      headers["Content-Type"] = "application/json";
+    }
+
+    let requestCompressionDuration;
+    let requestBody: Uint8Array | string | null = null;
+    if (body && compress && this.compression) {
+      headers["Content-Encoding"] = "gzip";
+      const beforeRequestCompression = performance.now();
+      requestBody = gzip(JSON.stringify(body));
+      requestCompressionDuration = performance.now() - beforeRequestCompression;
+    } else if (body) {
+      requestBody = JSON.stringify(body);
+    }
+
+    const maxAttempts = retryable ? 3 : 1;
+    let response!: Response;
+    let error: TurbopufferError | null = null;
+    let request_start!: number;
+    let response_start!: number;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      error = null;
+      request_start = performance.now();
+      try {
+        response = await fetch(new URL(path, this.origin), {
+          method,
+          headers,
+          body: requestBody,
+          signal: AbortSignal.timeout(this.connectTimeout),
+        });
+      } catch (e: unknown) {
+        if (e instanceof Error || e instanceof DOMException) {
+          if (e.cause instanceof Error) {
+            // wrap generic fetch failure errors with the underlying cause
+            error = new TurbopufferError(`fetch failed: ${e.cause.message}`, {
```

Comment:
do we still need this with hack with native fetch? it was a hack for undici specific behaviour

Reply:
only somewhat for DOMException since its output isn't too nice but I've since removed DOMException so this wrapping is no longer necessary

Reply:
Coming back to this... I've re-added the `TurbopufferError` wrapper for now. We'll have to do `error = new Error` or `error = new TurbopufferError` in DefaultHTTPClient.ts:
```
error = new TurbopufferError(
  message ?? `http error ${response.status}`,
  {
    status: response.status,
  },
);
```
I thought why not go with the latter; it's what NodeHTTPClient has been using and I think having the status code in the error output is nicer, which `TurbopufferError` already does. And so might as well also wrap the 'fetch failed' error to have all errors consistently wrapped!
(Having this sort of wrapper likely makes it easier to add more info to this output if needed in the future and at some point `TurbopufferError` can also be properly exported if/when users find it helpful.)